src/main/cpp/src/timezones.cu

```diff
@@ -118,6 +120,183 @@ auto convert_timestamp_tz(column_view const& input,
   return results;
 }
 
+template <typename timestamp_type>
+struct time_add_functor {
+  using duration_type = typename timestamp_type::duration;
+
+  lists_column_device_view const transitions;
+
+  size_type const tz_index;
+
+  int64_t const duration_scalar;
+
+  __device__ inline timestamp_type plus_with_tz(timestamp_type const& timestamp,
+                                                int64_t const& duration) const
+  {
+    if (duration == 0L) { return timestamp; }
+
+    auto const utc_instants = transitions.child().child(0);
+    auto const tz_instants  = transitions.child().child(1);
+    auto const utc_offsets  = transitions.child().child(2);
+
+    auto const epoch_seconds_utc = static_cast<int64_t>(
+      cuda::std::chrono::duration_cast<cudf::duration_s>(timestamp.time_since_epoch()).count());
+    // input epoch seconds
+
+    auto const tz_transitions = cudf::list_device_view{transitions, tz_index};
+    auto const list_size      = tz_transitions.size();
+
+    auto const transition_times_utc = cudf::device_span<int64_t const>(
+      utc_instants.data<int64_t>() + tz_transitions.element_offset(0),
+      static_cast<size_t>(list_size));
+
+    auto const transition_times_tz = cudf::device_span<int64_t const>(
+      tz_instants.data<int64_t>() + tz_transitions.element_offset(0),
+      static_cast<size_t>(list_size));
+
+    // step 1: Binary search on utc to find the correct offset to convert utc to local
+    auto const utc_it = thrust::upper_bound(
```

Comment:
Don't we do this some place else? Shouldn't transitioning between time zones like this be a static inline function? And if not can you add comments as to why this is different from `GpuTimeZoneDB.fromUtcTimestampToTimestamp`?